CenterForOpenScience/ember-osf-web
# [ENG-3726] Add all providers to files page

-   Ticket: [ENG-3726]
-   Feature flag: n/a

## Purpose

To allow the files page for projects to work properly with all providers. This includes turning off existing functionality for providers that do not support that functionality, and preparing for future functionality to be turned off.

## Summary of Changes

1. Add ProviderFiles and Files for each provider
2. Modify <provider>File and <provider>ProviderFile to turn off select functionality
3.  Only show Revisions/Tags panel if you can either see revisions or tags
4. Only show revisions button if the provider has meaningful revision information
5. Only show tags for OSF Storage
6. Increase the number of FileProviders we fetch from default to 20 (we have 12 providers)
7. Add flags providerHandlesVersioning, shouldShowRevisions, shouldShowTags, userCanDownloadAsZip, and userCanMoveToHere. 
8. Fix up currentUserPermissions
9. Make userCanMoveToHere especially useful for OSF Storage, checking for storage limits
10. Move OsfStorageManager to StorageManager and generalize for all providers
11. Selectively disable Download As Zip (especially for Github folders)
12. Disable folder creation and upload if no write permissions
13. Disable hashes in revisions tab if there are no hashes to show
14. Add back in Storage Providers to the moved section of the translations file
15. Fix tests
16. Add `target` to FileProvider Mirage
17. Allow deletes only when user has write permission
18. Disallow uploading to OSFStorage if over storage limit

## Screenshot(s)

Bitbucket:
<img width="1466" alt="bitbucket" src="https://user-images.githubusercontent.com/6599111/168165988-7d511e67-ed79-4761-a17d-5f1a6b808748.png">
<img width="1466" alt="bitbucket-file" src="https://user-images.githubusercontent.com/6599111/168166001-4b71d9c6-7a1e-4aae-b179-4e7d4012fc6e.png">

Box:
<img width="1466" alt="box" src="https://user-images.githubusercontent.com/6599111/168166034-3919c25e-368f-4b98-98d8-ca407e27febe.png">
<img width="1466" alt="box-file" src="https://user-images.githubusercontent.com/6599111/168166047-ceac6ebb-d32c-4577-a92c-ca1bb773152d.png">

Dataverse:
<img width="1466" alt="dataverse" src="https://user-images.githubusercontent.com/6599111/168166105-2f6fd990-a542-4ed3-ba4f-30834b2611f0.png">
<img width="1466" alt="dataverse-file" src="https://user-images.githubusercontent.com/6599111/168166123-edcfcff1-b3db-415a-a3e4-27c387713022.png">

Dropbox:
<img width="1422" alt="dropbox" src="https://user-images.githubusercontent.com/6599111/168166188-5b1bb781-3dc2-45b1-a0f4-9e6dc1600b9b.png">
<img width="1466" alt="dropbox-file" src="https://user-images.githubusercontent.com/6599111/168166202-78dbf2c3-29d8-4720-b383-167f7a476765.png">

Figshare:
<img width="1466" alt="figshare" src="https://user-images.githubusercontent.com/6599111/168166244-1945d901-f8bb-4908-a6f8-7390d7afbbea.png">
<img width="1466" alt="figshare-file" src="https://user-images.githubusercontent.com/6599111/168166253-c9f15498-86fc-455c-9052-aa9b20f9fd2c.png">

Github:
<img width="1466" alt="github" src="https://user-images.githubusercontent.com/6599111/168166291-d57278a2-91ad-4e89-b878-689496e1c75f.png">
<img width="1466" alt="github-file" src="https://user-images.githubusercontent.com/6599111/168166306-dc56560b-7087-4188-838f-173446ef183b.png">

Gitlab:
<img width="1466" alt="gitlab" src="https://user-images.githubusercontent.com/6599111/168166327-39704120-85e4-455c-adf7-d78f9fb528c2.png">
<img width="1466" alt="gitlab-file" src="https://user-images.githubusercontent.com/6599111/168166340-2a725c26-0fb3-41b4-b056-35ccefed0a0c.png">

Googledrive:
<img width="1466" alt="googledrive" src="https://user-images.githubusercontent.com/6599111/168166372-6254628d-a82b-48ba-a512-ab1967c3e169.png">
<img width="1466" alt="googledrive-file" src="https://user-images.githubusercontent.com/6599111/168166391-1c2a06de-7e15-49f9-94a6-4a5ab14c07db.png">

Onedrive:
<img width="1466" alt="onedrive" src="https://user-images.githubusercontent.com/6599111/168166413-6094d551-c900-4fe2-89da-290ebb66fb8e.png">
<img width="1466" alt="onedrive-file" src="https://user-images.githubusercontent.com/6599111/168166422-4e4a8c0e-a183-4d27-855b-2a0c345c9ea3.png">

OSFStorage:
<img width="1422" alt="osfstorage" src="https://user-images.githubusercontent.com/6599111/168166471-9a36ff4f-d7a2-4bcc-8eab-464fee181187.png">
<img width="1466" alt="osfstorage-file" src="https://user-images.githubusercontent.com/6599111/168166488-fa103fa2-6206-44fa-bb08-f6797a013f7f.png">

Owncloud:
<img width="1466" alt="owncloud" src="https://user-images.githubusercontent.com/6599111/168166524-e34a314b-54d4-4a10-92af-6e549589a3b6.png">
<img width="1466" alt="owncloud-file" src="https://user-images.githubusercontent.com/6599111/168166540-0b08c8b3-ad6c-43ff-b8e4-2dc1cc6c1dbb.png">

S3:
<img width="1422" alt="s3" src="https://user-images.githubusercontent.com/6599111/168166577-53ccc834-5e89-4514-a9fe-dc5a2bea5073.png">
<img width="1466" alt="s3-file" src="https://user-images.githubusercontent.com/6599111/168166601-50abd285-a48f-49a8-b57c-a586225fcf70.png">


## Side Effects

I hope not.

## QA Notes

Obviously, this complicates testing quite a bit because now there are potential combinations of things to test. The functionality should be what's indicated by product's chart of the capabilities of each provider, with the addition that according to a similar chart from Services, Github can't Download-as-Zip a folder, just the whole provider, and Dataverse and Figshare are now completely read-only.


[ENG-3726]: https://openscience.atlassian.net/browse/ENG-3726?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
